scenario/context.py

```diff
@@ -32,6 +43,99 @@ class InvalidActionError(InvalidEventError):
     """raised when something is wrong with the action passed to Context.run_action"""
 
 
+class AlreadyEmittedError(RuntimeError):
+    """Raised when _runner.run() is called more than once."""
+
+
+class _Manager:
+    """Context manager to offer test code some runtime charm object introspection."""
+
+    def __init__(
+        self,
+        ctx: "Context",
+        arg: Union[str, Action, Event],
+        state_in: "State",
+    ):
+        self._ctx = ctx
+        self._arg = arg
+        self._state_in = state_in
+
+        self._emitted: bool = False
+        self._run = None
+
+        self.charm: Optional[CharmType] = None
+        self.output: Optional[Union["State", ActionOutput]] = None
+
+    def _setup(self, charm: "CharmType"):
+        self.charm = charm
+
+    def _runner(self):
+        raise NotImplementedError("override in subclass")
+
+    def __enter__(self):
+        self._run = self._runner()
+        next(self._run)
+        return self
+
+    def run(self) -> "State":
+        """Emit the event and proceed with charm execution.
+
+        This can only be done once.
+        """
+        if self._emitted:
+            raise AlreadyEmittedError("Can only _runner.run() once.")
```

Comment:
`_runner` won't have meaning for the user. Can we just say "Can only call run() once"?